### DISCUSSION TITLE
Snowflake Notebooks

### DISCUSSION THREAD

**Discussion by BenKentzer** [littleK0i/SnowDDL#113] [General]
Hello!
Now that Notebooks are in public preview, is it possible to get the "create notebooks" grant added to SnowDDL? I'm currently manually allocating this permission to each user using global role, but it would be great if we could get to a point where notebooks permissions could be set from SnowDDL within the relevant schema.
Thanks
Ben

**Comment by littleK0i** [comment]
Notebooks feature still seems underbaked. For example, future grants for notebooks are not supported.

Anyway, I've added `NOTEBOOK` object type to version `0.30.3`. Now you may try to use it via permission models: https://docs.snowddl.com/basic/yaml-configs/permission-model

(1) Create new permission model and add CREATE NOTEBOOK grant:

```yaml
schema_with_notebooks:
  inherit_from: default

  owner_create_grants:
    - NOTEBOOK
```

(2) Use it from specific SCHEMA:

```yaml
permission_model: schema_with_notebooks
```

Since ownership is not set and cannot be set, notebooks will be "owned" by creator user roles. I suspect, notebooks will be executed with creator role permissions. Unimpressive, but not much we can do for now.

---

Btw, did you consider using Hex for notebooks instead?

https://hex.tech/

I suspect it will be much cheaper and easier to use & share compared to native notebooks.

**Reply by BenKentzer** [reply]
Thank you! I'm hoping that if / when Notebooks make it into GA, then the future grants and sharing issues will be resolved. 

We've looked at other notebooks providers (although not Hex) and are currently swayed by the simple set up and management of the Snowflake approach. Here's hoping it matures nicely...

Thanks
Ben

**Comment by BenKentzer** [comment]
Hello again!
It appears we have a bit of an issue.
When I create a brand new Snowflake account, notebooks work fine. I can create and execute them with no issues.
However, once a user is allocated a role that is set up by SnowDDL, it can create notebooks, but executing them times out.

I'm trying to find the "hidden" permission that is causing the problem, but not having any joy at the moment.

Any thoughts?

Thanks
Ben

**Reply by littleK0i** [reply]
Do you have notebook warehouse listed in "owner_warehouse_usage" for schema?
https://docs.snowddl.com/basic/yaml-configs/schema

**Reply by BenKentzer** [reply]
Hello!
I didn't, but I do now. 
I'm still getting the same problems - or at least, the same symptom - an undecipherable Snowflake error.
The warehouse starts, but the code won't run.
![image](https://github.com/user-attachments/assets/d674df12-8f02-4636-a0e8-9fe3082933a0)
Notebooks seem to be essentially a Streamlit app, so I've had a quick look at the permissions required for running a Streamlit. I'll continue on this track to see if I can find the culprit.

I've also noticed that when I run SnowDDL "plan" against an account after I've created a notebook, it suggests I run a command which will change the ownership to SnowDDL_Admin (we have this as our top-level SnowDDL deployment role). I have sandbox set to true:
![image](https://github.com/user-attachments/assets/4c1b8412-17cd-4e2c-8163-f7acb04bfcce)


Thanks
Ben

**Reply by littleK0i** [reply]
I see Snowflake created a technical warehouse of some sort: `SYSTEM$STREAMLIT_NOTEBOOK_WH`.

Maybe it is related somehow.

**Reply by BenKentzer** [reply]
Hello!

That's very interesting.

I've found this:
[https://docs.snowflake.com/en/user-guide/warehouses-overview#default-warehouse-for-notebooks](url)

It suggests that Snowflake creates this warehouse automatically for notebook usage. Interestingly, mine hasn't been started since it weas created. 
Usage should be set to PUBLIC so all users should be able to use it - at least, that's my assumption. Does SnowDDL do something clever with the PUBLIC role?

Thanks
Ben

**Reply by littleK0i** [reply]
Nothing. Public should be fully available.

We only drop "PUBLIC" schema which is automatically created for each database. But role should be fine.

Btw, you may see all executed commands with `--show-sql` CLI parameter.

**Reply by BenKentzer** [reply]
That's exactly as I expected (but just wanted to check).

That system warehouse doesn't get started on a different account where notebooks work either using the same python code. I'm just running the sample notebook that appears every time you create a new notebook.

I've re-raised it with Snowflake to see if they can see what's happening. Just waiting for my Sales Engineer to organise a time to walk through it.

Thanks

Ben

**Comment by BenKentzer** [comment]
![image](https://github.com/user-attachments/assets/2ee83f10-0369-4e4c-98c6-55f7a1bebffc)

Bit disaapointed that the guy managing the supprort ticket originally wasn't able to provide this information. I sent a video of the issue across which was downloaded at 8:45 this morning, and the SE sent the above email less than 15 minutes later.

Fingers crossed!

**Reply by BenKentzer** [reply]
Looks like the update still hasn't been deployed, so Notebooks are still not working properly :-(

**Comment by BenKentzer** [comment]
The change appears to have been deployed onto Snowflake now, so notebooks are all working as expected now :-)

Thanks for the support as always!
